nemo/collections/asr/parts/k2/topologies.py

```diff
@@ -0,0 +1,136 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import List
+
+import k2
+
+
+def build_topo(name: str, tokens: List[int], with_selfloops: bool = True) -> k2.Fsa:
+    """Helper function to build a topology.
+    Args:
+      name:
+        The topology name. Choices: default, compact, shared_blank, minimal
+      tokens:
+        A list of tokens, e.g., phones, characters, etc.
+      with_selfloops:
+        Whether to add token-to-epsilon self-loops to a topology
+    Returns:
+      Returns a topology FST.
+    """
+    if name == "default":
+        return build_default_topo(tokens, with_selfloops)
+    elif name == "compact":
+        return build_compact_topo(tokens, with_selfloops)
+    elif name == "shared_blank":
+        return build_shared_blank_topo(tokens, with_selfloops)
+    elif name == "minimal":
+        return build_minimal_topo(tokens)
+    else:
+        raise ValueError(f"Unknown topo name: {name}")
+
+
+def build_default_topo(tokens: List[int], with_selfloops: bool = True) -> k2.Fsa:
+    """Build the default CTC topology.
+    """
+    assert 0 in tokens, "We assume 0 is ID of the blank symbol"
+
+    num_states = len(tokens)
+    final_state = num_states
+    arcs = "" if with_selfloops else "0 0 0 0 0.0\n"
+    for i in range(num_states):
+        for j in range(num_states):
+            if i == j:
+                if with_selfloops:
+                    arcs += f"{i} {i} {tokens[i]} 0 0.0\n"
+            else:
+                arcs += f"{i} {j} {tokens[j]} {tokens[j]} 0.0\n"
+        arcs += f"{i} {final_state} -1 -1 0.0\n"
+    arcs += f"{final_state}"
+    ans = k2.Fsa.from_str(arcs, num_aux_labels=1)
+    ans = k2.arc_sort(ans)
+    return ans
+
+
+def build_compact_topo(tokens: List[int], with_selfloops: bool = True) -> k2.Fsa:
+    """Build the compact CTC topology.
+    See https://arxiv.org/abs/2110.03098
+    """
+    assert 0 in tokens, "We assume 0 is ID of the blank symbol"
+
+    selfloops_shift = int(with_selfloops)
+    blank_num = 1
+    num_states = len(tokens) + selfloops_shift
+    final_state = num_states
+    arcs = f"0 {selfloops_shift} {blank_num} 0 0.0\n"
+    for i in range(blank_num + selfloops_shift, num_states):
+        arcs += f"0 {i} {tokens[i - selfloops_shift] + 1} {tokens[i - selfloops_shift] + 1} 0.0\n"
+    arcs += f"0 {final_state} -1 -1 0.0\n"
+    for i in range(blank_num, num_states):
+        arcs += f"{i} 0 0 0 0.0\n"
+        if with_selfloops:
+            arcs += f"{i} {i} {tokens[i - selfloops_shift] + 1} 0 0.0\n"
+    arcs += f"{final_state}"
+    ans = k2.Fsa.from_str(arcs, num_aux_labels=1)
+    ans = k2.arc_sort(ans)
+    return ans
+
+
+def build_shared_blank_topo(tokens: List[int], with_selfloops: bool = True) -> k2.Fsa:
```

Comment:
Same as above

Reply:
Done in [7b2b4a4](https://github.com/NVIDIA/NeMo/pull/3351/commits/7b2b4a4ba4e34848d526d8c4ac53903bfbfe2fd2)